File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        
        if not digits:
            return []

        res = []
        graph = {
            2: ["a", "b", "c"],
            3: ["d", "e", "f"],
            4: ["g", "h", "i"],
            5: ["j", "k", "l"],
            6: ["m", "n", "o"],
            7: ["p", "q", "r", "s"],
            8: ["t", "u", "v"],
            9: ["w", "x", "y", "z"]
        }

        def backtrack(graph, index, curr):

            #Base Case
            if index == len(digits):
                res.append("".join(curr.copy()))
                return

            #Recursive
            for neighbor in graph[int(digits[index])]:
                curr.append(neighbor)
                backtrack(graph, index + 1, curr)
                curr.pop()
        

        backtrack(graph, 0, [])

        return res

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) 
    {

        if (digits.empty())
        {
            cout << "here" << endl;
            return res;
        }

        backtrack("", digits, 0);
        return res;
    }

    void backtrack(string curr, string& digits, int i)
    {

        //Base Case
        if (i == digits.size())
        {
            res.push_back(curr);
            return;
        }

        string letters = map[digits[i]];
        //Recursive
        for (char c : letters)
        {
            curr.push_back(c);
            backtrack(curr, digits, i + 1);
            curr.pop_back();
        }
    }

private:
    vector<string> res;
    unordered_map<char, string> map = 
    {
        {'2', "abc"},
        {'3', "def"},
        {'4', "ghi"},
        {'5', "jkl"},
        {'6', "mno"},
        {'7', "pqrs"},
        {'8', "tuv"},
        {'9', "wxyz"}
    };

};